src/graph/README.md

```diff
@@ -1,9 +1,206 @@
+
+## Validation and Deployment
+
+It is perhaps best to think of the controller as a compiler, except instead of taking and emitting code, it takes in yaml specifications via RampartGraph custom object and emits changes to the Kubernetes cluster.
+
+The core of this functionality is provided by the `Graph` class (found in `src/graph/objects/graph.py`). Of note, there are two asynchronous functions: `validate` and `deploy`.
+
+### Validation
+
+Validation occurs via four phases:
+
+1. Kubernetes rejects RampartGraph custom resources which are not valid yamls and do not meet the RampartGraph specification in the CRD. Currently, this is just that the graph contains `spec.graph.components`, `spec.graph.flows`, and an optional `spec.deploy`. This validation lives entirely outside of the Rampart Controller.
+2. The graph is validated against the JSONSchema defined in `src/graph/graph.json`. This occurs in the both the validator and controller.
+3. More complex static checks are performed. For example, references to flows in the components are checked for the correct type and mount paths are checked for validity.  This phase occurs in both the validator and controller.
+4. Checks that require pulling the component helm charts are performed. This includes that the helm charts are reachable and valid and that the annotations contained within the component charts match the specifications of the graph. Because this phase can timeout (say due to connection issues or to downloading a large helm chart), this phase is only performed in the controller. Because this phase cannot cause graphs that fail phase 1-3 to become valid, the validator will produce false positives, but will not produce false negatives. Therefore there will be graphs that are created that fail with `ValidationError`, but there will not be any correct graphs that fail to become created.
+
+Within the code itself, phase 2 is performed during instatiation of the `Graph` object by the classmethod `from_json`. Phases 3 and 4 are performed by the `validate` function, with phase 4 only performed when the argument `validate_specs` is set to `True`.
+
+### Deployment
+
+Once validation is completed, the controller can begin deployment of components and flows. To do this, it performs the following steps:
+
+1. Deploy all of the subnamespaces that will contain the components (note: each component gets its own namespace)
+2. Deploy all of the flows. Note that the actual body of volume flows do not exist per se, but are rather when two components using it write and read to the same place on a distributed filesystem
```

Comment:
In view mode, there is no line break after this sentence.
https://github.com/petuum/rampart/blob/0f1a29aaaa0c4aced83a3dbfdf0947c609cd9ad8/src/graph/README.md#deployment